src/site/content/en/blog/porting-libusb-to-webusb/index.md

```diff
@@ -0,0 +1,450 @@
+
+```js
+EM_JS(void, em_libusb_notify, (void), {
+    dispatchEvent(new Event("em-libusb"));
+});
+
+EM_ASYNC_JS(int, em_libusb_wait, (int timeout), {
+    let onEvent, timeoutId;
+
+    try {
+        return await new Promise(resolve => {
+            onEvent = () => resolve(0);
+            addEventListener('em-libusb', onEvent);
+
+            timeoutId = setTimeout(resolve, timeout, -1);
+        });
+    } finally {
+        removeEventListener('em-libusb', onEvent);
+        clearTimeout(timeoutId);
+    }
+});
+```
+
+The `em_libusb_notify` function is used whenever libusb tries to report an event, such as data transfer completion:
+
+```cpp
+void usbi_signal_event(usbi_event_t *event)
+{
+    uint64_t dummy = 1;
+    ssize_t r;
+
+    r = write(EVENT_WRITE_FD(event), &dummy, sizeof(dummy));
+    if (r != sizeof(dummy))
+        usbi_warn(NULL, "event write failed");
+#ifdef __EMSCRIPTEN__
+    em_libusb_notify();
+#endif
+}
+```
+
+Meanwhile, the `em_libusb_wait` part is used to "wake up" from Asyncify sleep when either an `em-libusb` event is received, or the timeout has expired:
+
+```cpp
+    double until_time = emscripten_get_now() + timeout_ms;
+    for (;;) {
+        // Emscripten `poll` ignores timeout param, but pass 0 explicitly just
+        // in case.
+        num_ready = poll(fds, nfds, 0);
+        if (num_ready != 0) break;
+        int timeout = until_time - emscripten_get_now();
+        if (timeout <= 0) break;
+        int result = em_libusb_wait(timeout);
+        if (result != 0) break;
+    }
+```
+
+Due to significant reduction in sleeps and wake-ups, this mechanism fixed the efficiency problems of the earlier `emscripten_sleep`-based implementation, and increased the DSLR demo throughput from 13-14 FPS to consistent 30+ FPS, which is enough for a smooth live feed.
+
+## Build system and the first test
+
+After the backend was done, I had to add it to `Makefile.am` and `configure.ac`. The only interesting bit here is Emscripten-specific flags modification:
+
+```bash
+emscripten)
+  AC_SUBST(EXEEXT, [.html])
+  # Note: LT_LDFLAGS is not enough here because we need link flags for executable.
+  AM_LDFLAGS="${AM_LDFLAGS} --bind -s ASYNCIFY -s ASSERTIONS -s ALLOW_MEMORY_GROWTH -s INVOKE_RUN=0 -s EXPORTED_RUNTIME_METHODS=['callMain']"
+  ;;
+```
+
+First, executables on Unix platforms normally don't have any extension. Emscripten, however, produces different output depending on which extension you request. I'm using `AC_SUBST(EXEEXT, …)` to change the executable extension to `.html` so that any executable within a package - tests and examples - becomes an HTML with Emscripten's default shell that takes care of loading and instantiating JavaScript and WebAssembly.
```

Comment:
```suggestion
First, executables on Unix platforms normally don't have a file extension. Emscripten, however, produces different output depending on which extension you request. I'm using `AC_SUBST(EXEEXT, …)` to change the executable extension to `.html` so that any executable within a package - tests and examples - becomes an HTML with Emscripten's default shell that takes care of loading and instantiating JavaScript and WebAssembly.
```

Reply:
Since it's "executables" (plural), should this be "file extensions" then?

Reply:
Yes. I fixed the suggestion.